Edited Aevol/doc/aevol_create.1
./"test with man -l <file>
.TH AEVOL "1" "July 2014" "aevol 4.4" "User Manual"
.SH NAME
aevol \- an in silico experimental evolution platform
.SH SYNOPSIS
.B aevol_create \-h
or
.B \-\-help
.br
.B aevol_create \-V
or
.B \-\-version
.br
.B aevol_create \fR[\fB\-v\fR] \fR[\fB\-f\fI PARAM_FILE\fR] \fR[\fB\-o\fI OUTDIR\fR] \fR[\fB\-c\fI CFILE\fR] \fR[\fB\-p\fI PFILE\fR]
.SH DESCRIPTION
.B Aevol
is a simulation platform that allows one to let populations of digital organisms evolve in different conditions and study experimentally the mechanisms responsible for the structuration of the genome and the transcriptome.
.TP
.B aevol_create
create an experiment with setup as specified in
.I param_file
.SH OPTIONS
.TP
.B \-h, \-\-help
print help, then exit
.TP
.B \-V, \-\-version
print version number, then exit
.HP
.B \-f, \-\-file
.I PARAM_FILE
.br
specify parameter file (default: param.in)
.HP
.B \-o, \-\-out
.I OUTDIR
.br
specify output directory (default "./")
.HP
.B \-c, \-\-chromosome
.I CFILE
.br
load chromosome from given text file instead of generating it
.HP
.B \-p, \-\-plasmid
.I PFILE
.br
load plasmid from given text file instead of generating it (params must allow plasmids)
.SH "SEE ALSO"
.B aevol_run, aevol_modify, aevol_propagate


Edited Aevol/ChangeLog
New in 8.0:
 - MPI Support:
    - Fully reproduceable MPI execution with sequential and memory shared parallel version (OpenMP)
    - Can be mixed up with OpenMP version: MPI+OpenMP
    - No (Explicit) Synchronization (i.e. no MPI Barrier). Should be very scalable.
    - RAevol (dynamic and static environment) and Aevol compatible
    - Lineage/Tree are fully supported
    
 - RAevol should be fixed
 - A lot of vectorization works on motif searchs function
 - Validated discrete fuzzy (against the continious fuzzy). WARNING : never compare results of discrete and continious fuzzy
 - A lot of memory leaks fixes
 - A lot of bug fixes
 - New differential search algorithm that does the diff for all metadatas (not only promoters all the metadatas Promoters/RNAs/Genes)

New in 4.4:

* aevol_create now gives a name to the created strain. It can be defined using
  the STRAIN_NAME keyword in the param file. If this keyword is not present, a
  default one will be generated.

* New option -o or --out in aevol_create
  Allows one to create the experiment in any location.

* aevol_propagate can now reset the PRNGs (random generators) with the new
  general option -S (affecting all random processes) or the new more specific
  options -m, -s, -t, -e and -n (each affects a single random process, see man
  page).

* The environmental target can be defined by a list of custom points using the
  ENV_ADD_POINT keyword in the parameter file. This is an alternative to the
  usual way, where it is defined as a sum of gaussians. Note, however, that
  environmental variation is possible only for the "gaussian" way.

* The aevol_misc_fixed_mutations program now outputs the number of genes
  affected by each mutation. More precisely, three new columns are produced:
    - the number of coding RNAs possibly disrupted by a switch, a small indel
      or a rearrangement breakpoint,
    - the number of coding RNAs entirely included in a rearranged segment,
    - in the case of a transfer by replacement, the number of genes that were
      entirely included in the replaced segment.
  This program does not work yet if plasmids are allowed.

* The programs aevol_misc_lineage, aevol_misc_ancstats,
  aevol_misc_fixed_mutations, aevol_misc_gene_families now work when the
  environment was changed at some point by aevol_modify.

* The fitness proportionate selection scheme now works with spatial structure

* In aevol_modify, it is now possible to replace a population by clones of the
  best individual. To do so, add the line CLONE_BEST_NO_TREE or
  CLONE_BEST_AND_CHANGE_TREE in the parameter file you supply to aevol_modify.
  Use CLONE_BEST_AND_CHANGE_TREE if your simulation campaign uses the
  genealogical trees, that is to say if you set RECORD_TREE to true. Otherwise,
  use CLONE_BEST_NO_TREE.

* Manual pages were added for aevol_misc_lineage, aevol_misc_ancstats,
  aevol_misc_fixed_mutations, aevol_misc_gene_families, aevol_misc_create_eps,
  aevol_misc_view_generation.

* New option (-c and -p) in aevol_create. Allows one to load a chromosome (and
  possibly plasmid) from a text file. If -p is used, ALLOW_PLASMIDS must be set
  to true and -c must also be used. If -c is used and ALLOW_PLASMIDS is set to
  true, -p must also be used.

* Add a script (src/misc/movies.py) to produce a movie from an aevol simulation
  with dumps activated.
  run ./src/misc/movies.py -h for help.
  Needs ffmpeg in path.

* The aevol_misc_gene_families program issues the detailed history of each gene
  family on the lineage of a given individual (providing its lineage file).
  A gene family is defined here as a set of coding sequences that arised by
  duplications of a single original gene. The original gene, called the root of
  the family, can either be one of the genes in the initial ancestor, or a new
  gene created from scratch (for example by a local mutation that transformed a
  non-coding RNA into a coding RNA). The history of gene duplications, gene
  losses and gene mutations in each gene family is represented by a binary tree.
  The program starts by loading the initial genome at the beginning of the
  lineage and by tagging each gene in this initial genome. Each of these initial
  genes is marked as the root of a gene family. Then, each mutation recorded in
  the lineage file is replayed and the fate of all tagged genes is followed and
  recorded in their respective families. When a gene is duplicated, the
  corresponding node in one of the gene trees becomes an internal node, and two
  children nodes are added to it, representing the two gene copies. When a gene
  sequence is modified, the mutation is recorded in its corresponding node in
  one of the gene trees. When a gene is lost, the corresponding node in one of
  the gene trees is labelled as lost. When a new gene appears from scratch, i.e.
  not by gene duplication, it becomes the root of a new gene tree. Environmental
  variations are also replayed exactly as they occured during the main run.
  When all mutations have been replayed, several output files are written in a
  directory called gene_trees. Two general text files are produced. The file
  called gene_tree_statistics.txt contains general data on each gene family,
  like its creation date, its extinction date, or how many nodes it contained.
  The file called nodeattr_tabular.txt contains information about each node of
  each gene tree, like when it was duplicated or lost or how many mutations
  occurred on its branch. In addition, for each gene tree, two text files are
  generated: a file called genetree******-topology.tre contains the topology of
  the gene tree in the Newick format, and a file called
  genetree******-nodeattr.txt that contains the detailed list of events that
  happened to each node in the tree file, before it was either duplicated or
  lost.

  Usage: ae_misc_gene_families [-c | -n] [-t tolerance] -f lineage_file


Changes in 4.4:


Edited Aevol/src/aevol_create.cpp
// ****************************************************************************
//
//          Aevol - An in silico experimental evolution platform
//
// ****************************************************************************
//
// Copyright: See the AUTHORS file provided with the package or <www.aevol.fr>
// Web: http://www.aevol.fr/
// E-mail: See <http://www.aevol.fr/contact/>
// Original Authors : Guillaume Beslon, Carole Knibbe, David Parsons
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// ****************************************************************************

/** \class
 *  \brief
 */

const char* DEFAULT_PARAM_FILE_NAME = "param.in";

// ============================================================================
//                                   Includes
// ============================================================================
#include <cstdlib>
#include <cstdio>
#include <cstring>
#include <assert.h>

#include <getopt.h>

#ifdef __X11
  #include "ExpManager_X11.h"
#else
  #include "ExpManager.h"
#endif

#include "ExpManager.h"
#include "ParamLoader.h"

#include "aevol_version.h"

using namespace aevol;

// Helper functions
void print_help(char* prog_path);
void interpret_cmd_line_options(int argc, char* argv[]);

// Command-line option variables
static char* param_file_name = nullptr;
static char* chromosome_file_name = nullptr;
static char* plasmid_file_name = nullptr;
static char* json_file_name = nullptr;
static ParamLoader* my_param_loader = nullptr;
static IOJson* my_iojson = nullptr;
static int nb_rank = 1;


int main(int argc, char* argv[]) {
  interpret_cmd_line_options(argc, argv);

  // Print the hash and date of the commit used to compile this executable
  printf("Running aevol version %s\n", version_string);

  if(param_file_name != nullptr) {
    // Create a param loader for the parameter file
    my_param_loader = new ParamLoader(param_file_name);
  }
  else if(json_file_name != nullptr) {
    my_iojson = new IOJson(json_file_name);
    assert(chromosome_file_name == nullptr || my_iojson->getNbrIndividuals() == 0);

  }

  // Initialize the experiment manager
  ExpManager* exp_manager = new ExpManager();


  // Initialize the simulation from the parameter file
  int32_t lchromosome = -1;
  char* chromosome = nullptr;

  if (chromosome_file_name != nullptr) {
    const int max_input_chrom_size = 1000000;
    char raw_chromosome[max_input_chrom_size];
    FILE* chromosome_file = fopen(chromosome_file_name, "r");
    if (chromosome_file == nullptr) {
      printf("ERROR: failed to open source chromosome file %s\n",
             chromosome_file_name);
      exit(EXIT_FAILURE);
    }
    if (fgets(raw_chromosome, max_input_chrom_size, chromosome_file) == nullptr)
    {
      printf("ERROR: failed to read from chromosome file %s\n",
             chromosome_file_name);
      exit(EXIT_FAILURE);
    }
    lchromosome = strlen(raw_chromosome) - 1;
    chromosome = new char[lchromosome]; // Warning: will become the DNA of the
                                        // first individual created -> do not
                                        // delete, will be freed in Dna
    strncpy(chromosome, raw_chromosome, lchromosome);
    printf("Loading chromosome from text file %s (%" PRId32 " base pairs) \n",
        chromosome_file_name, lchromosome);
    fclose(chromosome_file);
  }

  if(json_file_name != nullptr && my_iojson->getNbrIndividuals() != 0) {
    std::string str = my_iojson->getIndividualSequence(0,0);
    chromosome = new char[str.length()+1];
    strcpy(chromosome, str.c_str());
    lchromosome = strlen(chromosome);
  }

  int32_t lplasmid = -1;
  char* plasmid;

  if (plasmid_file_name != nullptr) {
    const int max_input_plasmid_size = 1000000;
    char raw_plasmid[max_input_plasmid_size];
    FILE* plasmid_file = fopen(plasmid_file_name, "r");
    if (plasmid_file == nullptr) {
      printf("ERROR: failed to open source chromosome file %s\n",
             plasmid_file_name);
      exit(EXIT_FAILURE);
    }
    if (fgets(raw_plasmid, max_input_plasmid_size, plasmid_file) == nullptr) {
      printf("ERROR: failed to read from chromosome file %s\n",
             chromosome_file_name);
      exit(EXIT_FAILURE);
    }
    lplasmid = strlen(raw_plasmid) - 1;
    plasmid = new char[lplasmid]; // Warning: will become the DNA of the first
                                  // individual created -> no not delete,
                                  // will be freed in Dna
    strncpy(plasmid, raw_plasmid, lplasmid);
    printf("Loading plasmid from text file %s (%" PRId32 " base pairs) \n",
        plasmid_file_name, lplasmid);
    fclose(plasmid_file);
  }

  if(param_file_name != nullptr) {
    if (lchromosome > -1) {
      if (lplasmid > -1) {
        my_param_loader->load(exp_manager, true, chromosome, lchromosome,
                              plasmid, lplasmid
                              #ifdef HAVE_MPI
                              , nb_rank
                              #endif
                              );
      } else {
        my_param_loader->load(exp_manager, true, chromosome, lchromosome
                              #ifdef HAVE_MPI
                              ,nullptr,0, nb_rank
                              #endif
                              );
      }
    } else {
      my_param_loader->load(exp_manager, true
                              #ifdef HAVE_MPI
                              ,nullptr,0,nullptr,0, nb_rank
                              #endif
                              );
    }
    delete my_param_loader;
  }
  if(json_file_name != nullptr) {
    if (lchromosome > -1) {
      if (lplasmid > -1) {
        my_iojson->load(exp_manager, true, chromosome, lchromosome,
                              plasmid, lplasmid);
      } else {
        my_iojson->load(exp_manager, true, chromosome, lchromosome);
      }
    } else {
      my_iojson->load(exp_manager, true);
    }
    delete my_iojson;
  }

  //~ ((ExpManager_X11*)exp_manager)->toggle_display_on_off();
  //~ exp_manager->display();
  //~ getchar();

  #ifdef HAVE_MPI
  exp_manager->set_nb_rank(nb_rank);
  #endif

  // 8) Save the experiment
  if(exp_manager->indivs().front()==nullptr){printf("null\n");}
  exp_manager->Save(true);

  delete exp_manager;
  delete[] param_file_name;
  delete[] json_file_name;
}







/*!
  \brief

*/
void print_help(char* prog_path) {
  // Get the program file-name in prog_name (strip prog_path of the path)
  char* prog_name; // No new, it will point to somewhere inside prog_path
  if ((prog_name = strrchr(prog_path, '/'))) {
    prog_name++;
  }
  else {
    prog_name = prog_path;
  }

  printf("******************************************************************************\n");
  printf("*                                                                            *\n");
  printf("*                        aevol - Artificial Evolution                        *\n");
  printf("*                                                                            *\n");
  printf("* Aevol is a simulation platform that allows one to let populations of       *\n");
  printf("* digital organisms evolve in different conditions and study experimentally  *\n");
  printf("* the mechanisms responsible for the structuration of the genome and the     *\n");
  printf("* transcriptome.                                                             *\n");
  printf("*                                                                            *\n");
  printf("******************************************************************************\n");
  printf("\n");
  printf("%s: create an experiment with setup as specified in PARAM_FILE.\n",
         prog_name);
  printf("\n");
  printf("Usage : %s -h or --help\n", prog_name);
  printf("   or : %s -V or --version\n", prog_name);
  printf("   or : %s [-f PARAM_FILE] [-C CHROM_FILE] [-P PLASMID_FILE]\n",
         prog_name);
  printf("\nOptions\n");
  printf("  -h, --help\n\tprint this help, then exit\n");
  printf("  -V, --version\n\tprint version number, then exit\n");
  printf("  -f, --file PARAM_FILE\n");
  printf("\tspecify parameter file (default: param.in)\n");
  printf("  -C, --chromosome CHROM_FILE\n");
  printf("\tload chromosome from given text file instead of generating it\n");
  printf("  -P, --plasmid PLASMID_FILE\n");
  printf("\tload plasmid from given text file instead of generating it\n");
  #ifdef HAVE_MPI
  printf("  -r, --rank NB_RANK\n");
  printf("\tnumber of MPI ranks to declare\n");
  #endif
}

void interpret_cmd_line_options(int argc, char* argv[]) {
  // Define allowed options
  const char* options_list = "hVf:C:P:J:r:";
  static struct option long_options_list[] = {
      {"help",       no_argument,       nullptr, 'h'},
      {"version",    no_argument,       nullptr, 'V'},
      {"file",       required_argument, nullptr, 'f'},
      {"chromosome", required_argument, nullptr, 'C'},
      {"plasmid",    required_argument, nullptr, 'P'},
      {"json",       required_argument, nullptr, 'J'},
      #ifdef HAVE_MPI
      {"rank",       required_argument, nullptr, 'r'},
      #endif
      {0, 0, 0, 0}
  };

  // Get actual values of the CLI options
  int option;
  while ((option = getopt_long(argc, argv, options_list, long_options_list,
                               nullptr)) != -1) {
    switch (option) {
      case 'h' : {
        print_help(argv[0]);
        exit(EXIT_SUCCESS);
      }
      case 'V' : {
        Utils::PrintAevolVersion();
        exit(EXIT_SUCCESS);
      }
      case 'f' : {
        param_file_name = new char[strlen(optarg) + 1];
        strcpy(param_file_name, optarg);
        break;
      }
      case 'C': {
        chromosome_file_name = new char[strlen(optarg) + 1];
        strcpy(chromosome_file_name, optarg);
        break;
      }
      case 'P': {
        plasmid_file_name = new char[strlen(optarg) + 1];
        strcpy(plasmid_file_name, optarg);
        break;
      }
      case 'J': {
        json_file_name = new char[strlen(optarg) + 1];
        strcpy(json_file_name, optarg);
        break;
      }
      #ifdef HAVE_MPI
      case 'r':  {
        nb_rank = atol(optarg);
        break;
      }
      #endif
      default : {
        printf("error %s %d\n",optarg,option);
        // An error message is printed in getopt_long, we just need to exit
        exit(EXIT_FAILURE);
      }
    }
  }
  assert(param_file_name == nullptr || json_file_name == nullptr);

  // Set undefined command line parameters to default values
  if (param_file_name == nullptr && json_file_name == nullptr) {
    param_file_name = new char[strlen(DEFAULT_PARAM_FILE_NAME) + 1];
    sprintf(param_file_name, "%s", DEFAULT_PARAM_FILE_NAME);
  }
}
